File: tests/fake_mcp_server.mjs
// Tiny stdio MCP server for the Node mcp-client tests (paginates tools/list).
import { createInterface } from "node:readline";

function send(obj) {
  process.stdout.write(JSON.stringify(obj) + "\n");
}

createInterface({ input: process.stdin }).on("line", (line) => {
  let msg;
  try {
    msg = JSON.parse(line);
  } catch {
    return;
  }
  if (msg.method === "initialize") {
    send({
      jsonrpc: "2.0",
      id: msg.id,
      result: {
        protocolVersion: "2024-11-05",
        capabilities: {},
        serverInfo: { name: "fake", version: "0" },
      },
    });
  } else if (msg.method === "tools/list") {
    const cursor = (msg.params || {}).cursor;
    if (!cursor) {
      send({
        jsonrpc: "2.0",
        id: msg.id,
        result: {
          tools: [{ name: "alpha", description: "a", inputSchema: { type: "object", properties: {} } }],
          nextCursor: "c1",
        },
      });
    } else {
      send({ jsonrpc: "2.0", id: msg.id, result: { tools: [{ name: "beta", description: "b" }] } });
    }
  } else if (msg.method === "tools/call") {
    const name = (msg.params || {}).name;
    if (name === "boom") {
      send({ jsonrpc: "2.0", id: msg.id, result: { isError: true, content: [{ type: "text", text: "nope" }] } });
    } else {
      send({ jsonrpc: "2.0", id: msg.id, result: { content: [{ type: "text", text: "ok:" + name }] } });
    }
  }
});

File: tests/fake_mcp_server.py
"""Tiny stdio MCP server used by the host-side mcp.py tests (no deps)."""

import json
import sys


def send(obj):
    sys.stdout.write(json.dumps(obj) + "\n")
    sys.stdout.flush()


for line in sys.stdin:
    line = line.strip()
    if not line:
        continue
    try:
        msg = json.loads(line)
    except json.JSONDecodeError:
        continue
    method = msg.get("method")
    if method == "initialize":
        send({"jsonrpc": "2.0", "id": msg["id"], "result": {
            "protocolVersion": "2024-11-05",
            "capabilities": {},
            "serverInfo": {"name": "fake", "version": "0"},
        }})
    elif method == "tools/list":
        send({"jsonrpc": "2.0", "id": msg["id"], "result": {"tools": [
            {"name": "alpha", "description": "a", "inputSchema": {"type": "object", "properties": {}}},
            {"name": "beta", "description": "b", "inputSchema": {"type": "object", "properties": {}}},
        ]}})
    elif method == "tools/call":
        send({"jsonrpc": "2.0", "id": msg["id"], "result": {
            "content": [{"type": "text", "text": "ok"}],
        }})
